packages/gatsby-plugin-sharp/src/image-data.ts

```diff
@@ -0,0 +1,230 @@
+
+export async function getImageMetadata(
+  file: FileNode,
+  getDominantColor?: boolean
+): Promise<IImageMetadata> {
+  if (!getDominantColor) {
+    // If we don't need the dominant color we can use the cheaper size function
+    const { width, height, type } = await getImageSizeAsync(file)
+    return { width, height, format: type }
+  }
+  let metadata = metadataCache.get(file.internal.contentDigest)
+  if (metadata && process.env.NODE_ENV !== `test`) {
+    return metadata
+  }
+  const pipeline: SharpInstance = sharp(file.absolutePath)
+
+  const { width, height, density, format } = await pipeline.metadata()
+
+  const { dominant } = await pipeline.stats()
+  // Fallback in case sharp doesn't support dominant
+  const dominantColor = dominant
+    ? rgbToHex(dominant.r, dominant.g, dominant.b)
+    : `#000000`
+
+  metadata = { width, height, density, format, dominantColor }
+  metadataCache.set(file.internal.contentDigest, metadata)
+  return metadata
+}
+
+export interface IImageDataProps {
+  file: FileNode
+  args: ISharpGatsbyImageArgs
+  cache: GatsbyCache
+  reporter: Reporter
+}
+
+export interface IImageDataArgs {
+  file: FileNode
+  args: ISharpGatsbyImageArgs
+  pathPrefix: string
+  cache: GatsbyCache
+  reporter: Reporter
+}
+
+export async function generateImageData({
+  file,
+  args,
+  pathPrefix,
+  reporter,
+  cache,
+}: IImageDataArgs): Promise<ISharpGatsbyImageData | undefined> {
+  const {
+    layout = `fixed`,
+    placeholder = `dominantColor`,
+    tracedSVGOptions = {},
+  } = args
+
+  const metadata = await getImageMetadata(file, placeholder === `dominantColor`)
+
+  const imageSizes: {
+    sizes: Array<number>
+    presentationWidth: number
+    presentationHeight: number
+    aspectRatio: number
+    isTopSizeOverriden: boolean
+  } = calculateImageSizes({
+    file,
+    layout,
+    ...args,
+    imgDimensions: metadata,
+    reporter,
+  })
+
+  const transforms = imageSizes.sizes.map(outputWidth => {
+    const width = Math.round(outputWidth)
+    const transform = createTransformObject({
+      ...args,
+      width,
+      height: Math.round(width / imageSizes.aspectRatio),
+      toFormat: args.toFormat || metadata.format,
+    })
+
+    if (pathPrefix) transform.pathPrefix = pathPrefix
+    return transform
+  })
+
+  const images = batchQueueImageResizing({
+    file,
+    transforms,
+    reporter,
+  })
+
+  const srcSet = getSrcSet(images)
+
+  const widthOfMaxSize = imageSizes.isTopSizeOverriden
+    ? metadata.width
+    : imageSizes.presentationWidth
+
+  const sizes = args.sizes || getSizes(widthOfMaxSize, layout)
+
+  const primaryIndex = imageSizes.sizes.findIndex(
+    size => size === widthOfMaxSize
+  )
+
+  if (primaryIndex === -1) {
+    reporter.panic(
+      `No image of the specified size found. Images may not have been processed correctly.`
+    )
+  }
+
+  const primaryImage = images[primaryIndex]
+
+  if (!images?.length) {
```

Comment:
images should always be a [] unsure why ou need the ?
```suggestion
  if (!images.length) {
```